modules/gapi/src/backends/fluid/gfluidcore_func.simd.hpp

```diff
@@ -292,279 +337,241 @@ CV_ALWAYS_INLINE void v_store_select(short* dst, const v_int16& div, const v_int
+    v_float32 fdiv1 = vg_load_f32(in2x);
+    v_float32 fdiv2 = vg_load_f32(&in2x[nlanes / 2]);
 
-    v_float32 scale = vx_setall_f32(static_cast<float>(_scale));
-    v_float32 v_zero = vx_setall_f32(0);
-    int x = 0;
-    for (;;)
-    {
-        for (; x <= length - nlanes; x += nlanes)
-        {
-            v_float32 a1 = vg_load_f32(&in1[x]);
-            v_float32 a2 = vg_load_f32(&in1[x + nlanes / 2]);
+    v_float32 r1 = div_op(s_tag, a1, fdiv1, v_scale);
+    v_float32 r2 = div_op(s_tag, a2, fdiv2, v_scale);
 
-            v_float32 fdiv1 = vg_load_f32(&in2[x]);
-            v_float32 fdiv2 = vg_load_f32(&in2[x + nlanes / 2]);
+    v_int32 res1 = v_round(v_select((fdiv1 == v_zero), v_zero, r1));
+    v_int32 res2 = v_round(v_select((fdiv2 == v_zero), v_zero, r2));
 
-            v_float32 r1 = div_op(t, a1, fdiv1, scale);
-            v_float32 r2 = div_op(t, a2, fdiv2, scale);
+    v_store_i16(outx, res1, res2);
+}
 
-            v_int32 res1 = v_round(v_select((fdiv1 == v_zero), v_zero, r1));
-            v_int32 res2 = v_round(v_select((fdiv2 == v_zero), v_zero, r2));
+template<typename scale_tag_t, typename SRC, typename DST, typename Vtype>
+CV_ALWAYS_INLINE
+typename std::enable_if<std::is_same<DST, short>::value ||
+                        std::is_same<DST, ushort>::value, void>::type
+div_hal(scale_tag_t s_tag, const SRC* in1x, const SRC* in2x, DST* outx,
+        const v_float32& v_scale, const Vtype& v_zero)
+{
+    constexpr int nlanes = vector_type_of_t<DST>::nlanes;
 
-            v_store_i16(&out[x], res1, res2);
-        }
+    v_float32 a1 = vg_load_f32(in1x);
+    v_float32 a2 = vg_load_f32(&in1x[nlanes / 2]);
 
-        if (x < length)
-        {
-            x = length - nlanes;
-            continue;  // process one more time (unaligned tail)
-        }
-        break;
-    }
-    return x;
+    div_simd_impl(s_tag, a1, a2, in2x, outx, v_scale, v_zero);
 }
 
 //-------------------------------------------------------------------------------------------------
 
 template<typename scale_tag_t, typename SRC>
-CV_ALWAYS_INLINE int div_hal(scale_tag_t t, const SRC in1[], const SRC in2[], float out[],
-                             const int length, double _scale)
+CV_ALWAYS_INLINE void div_simd_impl(scale_tag_t s_tag, const v_float32& a1, const SRC* in2x,
+                                           float* outx, const v_float32& v_scale)
 {
-    constexpr int nlanes = v_float32::nlanes;
+    v_float32 b1 = vg_load_f32(in2x);
+    vx_store(outx, div_op(s_tag, a1, b1, v_scale));
+}
+
+template<typename scale_tag_t, typename SRC, typename Tvec>
+CV_ALWAYS_INLINE void div_hal(scale_tag_t s_tag, const SRC* in1x, const SRC* in2x, float* outx,
+                              const v_float32& v_scale, const Tvec&)
+{
+    v_float32 a1 = vg_load_f32(in1x);
+    div_simd_impl(s_tag, a1, in2x, outx, v_scale);
+}
+
+//-------------------------------------------------------------------------------------------------
+#if 0
```

Comment:
Do we need this? If yes please put a comment why

Reply:
Applied.